Adding index, create, show, update endpoints for categories

green-arrow opened the pull request:
Addresses #27

green-arrow committed:
Adding index, create, show, update endpoints for categories

joshsmith commented on priv/repo/migrations/20160808143454_create_category.exs:
```diff
@@ -0,0 +1,15 @@
+      timestamps()
+    end
+
+    create index(:categories, [:slug], unique: true, name: :index_categories_on_slug)
```
Is this case-insensitive? I think you may need a `lower` here.

green-arrow replied:
I didn't notice that in the schema / migrations in the ruby project, so I didn't add it here. Does ruby do that by default?

Also, I believe that if we create the index via `lower`, we'll have to make sure that whenever we query for the slug we always lower case it in the query. If that's correct I'll change it here :)

joshsmith commented on test/controllers/category_controller_test.exs:
```diff
@@ -0,0 +1,86 @@
+
+  alias CodeCorps.Category
+
+  @valid_attrs %{description: "The technology category", name: "Technology", slug: "technology"}
```
Would reword as:

> You want to improve software tools and infrastructure.

For `description`.

joshsmith commented on test/models/category_test.exs:
```diff
@@ -0,0 +1,18 @@
+
+  alias CodeCorps.Category
+
+  @valid_attrs %{description: "The technology category", name: "Technology", slug: "technology"}
```
Would also rename `description` here as above.

joshsmith commented:
Looks great!

One thing we should probably change is not allowing `slug` to be set directly but instead having `slug` set on creation by using `Inflex.parameterize/1`, as in the private `generate_slug/1` in the `Organization` model.

joshsmith commented:
Another thing of note (not handled in previous endpoints) is that the create/update endpoints should only be allowed for admin users. There's no concept right now of an admin user.

green-arrow commented:
@JoshSmith how many times are generated slugs used in the database? If it's more than just here and in the organizations, I'm going to abstract out that `generate_slug/1` method to a model helper module.

joshsmith commented:
Similarly, it might be nice to bring over the seeds we had in Rails. Can we use such seeds in production, too?

``` ruby
categories = [
  {
    name: "Arts",
    description: "You want to improve the arts."
  },
  {
    name: "Economy",
    description: "You want to improve finance and the economic climate."
  },
  {
    name: "Education",
    description: "You want to improve literacy, schools, and training."
  },
  {
    name: "Environment",
    description: "You want to improve your environment."
  },
  {
    name: "Government",
    description: "You want to improve government responsiveness."
  },
  {
    name: "Health",
    description: "You want to improve prevention and treatment."
  },
  {
    name: "Justice",
    description: "You want to improve your judicial system."
  },
  {
    name: "Politics",
    description: "You want to improve elections and voting."
  },
  {
    name: "Public Safety",
    description: "You want to improve crime prevention and safety."
  },
  {
    name: "Science",
    description: "You want to improve tools for advancing science."
  },
  {
    name: "Security",
    description: "You want to improve tools like encryption."
  },
  {
    name: "Society",
    description: "You want to improve our communities."
  },
  {
    name: "Technology",
    description: "You want to improve software tools and infrastructure."
  },
  {
    name: "Transportation",
    description: "You want to improve how people travel."
  },
]
```

joshsmith commented:
@green-arrow go ahead and abstract it out. I'm not positive on where else it's used, but you can assume (for SEO purposes) that it may be reused often.

green-arrow committed:
Updates based on code review comments.

Changed description of technology category in tests.
Added automatic slug generation for categories.
Changed category slug index to be case-insensitive.
Added category seed data (just run 'mix run priv/repo/seeds.exs')

green-arrow commented:
@JoshSmith I think I addressed all comments in the last commit. Let me know if you see anything else that needs changed :)

joshsmith commented:
This looks good to me. Merge when ready. 👍
